Voice instructions caching using ResourceLoader.

tomaszrybakiewicz opened the pull request:
Closes [#896](https://github.com/mapbox/navigation-sdks/issues/896)

### Description

- Updated `libnavui-voice` to load audio data using `libnavui-util` `ResourceLoader`.
- Simplified audio loading logic by removing overly complex `VoiceActions`, `VoiceResult`, and `VoiceProcessor` classes.
- Removed `com.mapbox.mapboxsdk:mapbox-sdk-services` dependency and all `okhttp`, and `retrofit` references from `libnavui-voice` module.
- Added `com.mapbox.mapboxsdk:mapbox-sdk-services` dependency to `libnavui-dropin` module (before it was implicitly used via `libnavui-voice` module)

| Before | After |
|:---:|:---:|
|![navSDK 2 4 0 assets loading - voice](https://user-images.githubusercontent.com/2678039/167441469-b3ad18c0-245a-476d-8129-db4a1f797bac.png) | ![navSDK 2 4 0 assets loading - using TileStore rev4  - voice](https://user-images.githubusercontent.com/2678039/167450145-f76dc909-87bd-4b51-a904-0758a1fb2dc2.png)|

tomaszrybakiewicz commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/UrlUtils.kt:
```diff
@@ -0,0 +1,167 @@
+/*
```
@LukasPaczos Not sure if this is the right way to embed other sources in our codebase.

The previous implementation was URL encoding "ssml" voice instructions in a specific way (see _String.canonicalize_ method below) that `api.mapbox.com` accepts. Unfortunately, that encoding was done implicitly by a utility class in okhttp3 lib that's used by `com.mapbox.mapboxsdk:mapbox-sdk-services`.

kmadsen commented on libnavui-voice/build.gradle:
```diff
@@ -41,8 +41,6 @@ dependencies {
     api project(":libnavui-base")
     implementation project(":libnavui-resources")
 
-    api dependenciesList.mapboxSdkServices
```
This will potentially break downstream upgrades which is essentially a SEMVER 🤔 . Why is this being moved?

tomaszrybakiewicz replied:
`libnavui-voice` doesn't need this dependency anymore.

The modules that implicitly referenced `mapboxSdkServices` classes via `libnavui-voice` are: `libnavui-dropin` and `examples`.

kmadsen replied:
Yeah `examples` represents a customer, so this will be a breaking change if are using `mapboxSdkServices`. The customer will not be able to simply, upgrade a minor version, they will need to add a dependency if they are depending on `mapboxSdkServices`

Seems we may want to add `api dependenciesList.mapboxSdkServices` to `libnavigation-android` to keep it as part of the sdk bundle 🤔

LukasPaczos replied:
`libnavui-voice` can be used independently of any other module, so removing a dependency that has been made public might in some circumstances break compilation downstream.

I believe that after this PR we don't have any dependencies on the HTTP services from mapbox-java so we could only rely on the models packages and decrease the binary size of the whole SDK. We could potentially take that decision (and document it appropriately) but let's discuss it separately from this PR. For now, let's leave the public dependency as-is.

LukasPaczos commented on libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechApiTest.kt:
```diff
@@ -116,152 +109,152 @@ class MapboxSpeechApiTest {
-                exceptions[0].localizedMessage
-            )
-        }
+    // @Test
```
Let's clean this up.

LukasPaczos reviewed (commented):
Currently the TTS fallback does not work. After disabling internet connection, when the `VoiceInstructionsObserver` fires the messages seem to get queued up and errors in `speechCallback` are never delivered so that we can fallback to the onboard TTS player. After re-enabling the internet connection, all the queued up messages are downloaded and delivered which looks like it's either somewhere in our error handling or even in the Common SDK. I haven't debugged the code to see where exactly the queue forms.

Let's also add a test to ensure that this basic use-case is verified by the automation, possibly an instrumentation test might work best here to go through the happy path.

tomaszrybakiewicz commented:
@LukasPaczos I've applied your suggestions, patched DefaultResourceLoader to handle network reachability, updated the LICENSE file, and added missing voice instructions "fallback" tests. Please let me know if there is anything else that needs to change.

LukasPaczos commented on libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/resource/DefaultResourceLoader.kt:
```diff
@@ -43,20 +48,33 @@ internal class DefaultResourceLoader(
-        ) {
-            cancelableMap.remove(requestId)
-            callbackAdapter.run(it)
+        // Since the TileStore (commonSDK 21.3.1) will defer any requests that require internet connection,
+        // we must verify network requirement here and fail fast if the network is not available.
+        if (requiresNetwork && !reachability.isReachable) {
+            callbackAdapter.run(createError(connectionError()))
```
This is too board of a check. If we're offline, we still want to check the cache and see whether the resource is available and only fail if it isn't (so that we can fallback to TTS). It looks like we'll need an additional change from Common SDK in form of a flag to fail resource loading if cache is empty and there's no internet connection.

I wouldn't say it's a blocker for the PR as a whole but it would have to be fixed before we call the feature complete.

tomaszrybakiewicz replied:
When we are offline, we could try and load the cached resource by calling the `TileStore` with `ResourceLoadFlags.ACCEPT_EXPIRED` and `NetworkRestriction.DISALLOW_ALL` flags. The `TileStore` would then return only the available cached resource without accessing the network.

LukasPaczos replied:
Good idea! I would still hope that an SDK wouldn't have to check the connectivity status when dealing with a resource provider, but this approach sounds like a nice stop-gap. Let's do that and at the same time cut a Common SDK feature request to track for the future.

tomaszrybakiewicz replied:
Unfortunately, the  [suggested fix above](https://github.com/mapbox/mapbox-navigation-android/pull/5790#discussion_r875907681) doesn't work. 

For some reason, when in offline mode, calling the `TilesStore::loadResource` with `ResourceLoadFlags.ACCEPT_EXPIRED` and `NetworkRestriction.DISALLOW_ALL` flags doesn't tigger callback with `Unsatisfied` error  as expected (according to [source](https://github.com/mapbox/mapbox-sdk-common/blob/v21.3.1/components/tile_store/src/mapbox/common/tile_store_worker.cpp#L700)).

Instead, the `TileStore` logs an error to the console.

```
W/Mapbox: [tile_store] id=1344, tag=DefaultResourceLoader-6, resource=navigation:https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20300%20meters,%20Turn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3ELake%20Shore%20Boulevard%20West%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10apWNHMx4JS4: failed with [NetworkError] Error happened during okhttp download session: Not allowed: [ConnectionError] Not allowed; delaying for 1s after 1 attempts; 1 downloads waiting
```

I think we should merge this PR as is and cut the follow-up ticket for updating the DefaultResourceLoader implementation later once Common SDK has been updated.

tomaszrybakiewicz replied:
Logged this issue in CommonSDK repo https://github.com/mapbox/mapbox-sdk-common/issues/2744

abhishek1508 replied:
Now that this suggested stop-gap solution does work, 
> Good idea! I would still hope that an SDK wouldn't have to check the connectivity status when dealing with a resource provider, but this approach sounds like a nice stop-gap.Let's do that and at the same time cut a Common SDK feature request to track for the future.

This still holds true. Did we already create an improvement ticket for this on Common SDK side? @tomaszrybakiewicz

LukasPaczos reviewed (commented):
I took this for another spin and we seem to be constantly missing the cache. Replaying the same section of the route, with the same instruction, I still see network traffic to download the mp3 file.

For example, first run:
```
https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPowsta%C5%84c%C3%B3w%20%C5%9Al%C4%85skich%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=<TOKEN>>&sku=10aqALRA6wa0J
```
and second run:
```
https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPowsta%C5%84c%C3%B3w%20%C5%9Al%C4%85skich%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=<TOKEN>>&sku=10aJjhSB67Ns9
```
the only difference in the URL being the SKU token.

Are there any logs we can look for to debug this @anderco @zach2good @kkaefer? What's the database in which the blobs should be stored? I don't see anything new being created on disk.

anderco commented:
> I took this for another spin and we seem to be constantly missing the cache. Replaying the same section of the route, with the same instruction, I still see network traffic to download the mp3 file.
> 
> For example, first run:
> 
> ```
> https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPowsta%C5%84c%C3%B3w%20%C5%9Al%C4%85skich%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=<TOKEN>>&sku=10aqALRA6wa0J
> ```
> 
> and second run:
> 
> ```
> https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPowsta%C5%84c%C3%B3w%20%C5%9Al%C4%85skich%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=<TOKEN>>&sku=10aJjhSB67Ns9
> ```
> 
> the only difference in the URL being the SKU token.

Which version of Common SDK are you using? This shouldn't be  a problem since [v22.0.0-beta.1](https://github.com/mapbox/mapbox-sdk-common/pull/2666).

> Are there any logs we can look for to debug this @anderco @zach2good @kkaefer? What's the database in which the blobs should be stored? I don't see anything new being created on disk.

There should be more detail in the logs if you set the "log-tile-loading" tile store option to true. The blobs are stored as individual files on disk in the `resources` directory inside TileStore's root path. Metadata should be in the same `metadata.db` database used for tiles.

LukasPaczos commented:
We are on Common SDK `v21.3.1`, this means we need to wait until an upgrade to `v22.0.0-beta.1` lands next week.

That said, I still don't see any `resources` directory, continuing in https://github.com/mapbox/mapbox-sdk-common/issues/2733.

LukasPaczos commented:
> That said, I still don't see any resources directory

Capturing from the above that the actual path is:
```
CoreInitializer.createSystemInformation().applicationDataPath + "/.mapbox" + "/tile_store"
```
making it hidden.

Let's wait for Common SDK `v22.0.0-beta.1` integration and then we can continue with this PR.

LukasPaczos commented:
> That said, I still don't see any resources directory [...]. Capturing from the above that the actual path is [...] hidden [and not visible in the file explorer].

This is a bug just fixed in the latest Android Studio Chipmunk - https://issuetracker.google.com/issues/140057005!

@tomaszrybakiewicz could you rebase on top of `main` which has the updated Common SDK version?

tomaszrybakiewicz commented:
> Let's wait for Common SDK v22.0.0-beta.1 integration and then we can continue with this PR.

@LukasPaczos unfortunately, the SKU issue is still present in Common SDK `v22.0.0-beta.1`.

Take a look at the logs for two nav sessions that use the same voice instructions. Just like before, only the SKU value is different.

<details><summary>Expand for logs</summary>

```
------------------- Session 1 ------------------------
D/Mapbox: [nav-native]: Directions Service: Get route request: https://api.mapbox.com/directions/v5/mapbox/driving-traffic/-79.4850152,43.6245377;-79.4873091,43.6298485?access_token=REDACTED&geometries=polyline6&overview=full&steps=true&layers=0%3B&continue_straight=true&annotations=congestion_numeric%2Cmaxspeed%2Cclosure%2Cspeed%2Cduration%2Cdistance&language=en&roundabout_exits=true&voice_instructions=true&banner_instructions=true&voice_units=metric&enable_refresh=true

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20northeast.%20Then%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPark%20Lawn%20Road%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20northeast.%20Then%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPark%20Lawn%20Road%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20600%20meters,%20Turn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20600%20meters,%20Turn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aRXFKDG8AVD

...

------------------- Session 2 ------------------------
D/Mapbox: [nav-native]: Directions Service: Get route request: https://api.mapbox.com/directions/v5/mapbox/driving-traffic/-79.4850152,43.6245377;-79.4872975,43.629709?access_token=REDACTED&geometries=polyline6&overview=full&steps=true&layers=0%3B&continue_straight=true&annotations=congestion_numeric%2Cmaxspeed%2Cclosure%2Cspeed%2Cduration%2Cdistance&language=en&roundabout_exits=true&voice_instructions=true&banner_instructions=true&voice_units=metric&enable_refresh=true

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20northeast.%20Then%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPark%20Lawn%20Road%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20northeast.%20Then%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPark%20Lawn%20Road%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20600%20meters,%20Turn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20600%20meters,%20Turn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU

D/Mapbox: [nav-sdk]: [LoggingInterceptor] >> onDownload https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU|{x-accept-encoding=br, gzip, if-none-match=, user-agent=com.mapbox.navigation.examples/v2.5.0-beta.2-59-ge078235/9 MapboxCommon/22.0.0-beta.1 Android (32; oriole (Pixel 6))}
D/Mapbox: [nav-sdk]: [LoggingInterceptor] << onResponse 200 https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aQOZzNH39XU

...
```
</details>

Some implementation details: 
The `MapboxSpeechProvider` is using `MapboxNavigationAccounts.obtainUrlWithSkuToken` which behind the scenes calls  `BillingServiceInterface.getSessionSKUTokenIfValid` native method from Common SDK to acquire SKU token. 
The method name implies that the SKU token is for the current "session" only. 

Does that mean it will change for each session?

abhishek1508 commented:
The fix from Common will be available in `v22.0.0`

tomaszrybakiewicz commented:
I can confirm that the fix introduced in Common SDK `v22.0.0` does work.

Here are the logs captured from ResourceLoader observer. 
`transferredBytes=0` indicates that the resource has been loaded from the DISK CACHE instead of the NETWORK.

```
---- (route start) ---
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=18696, transferredBytes=0, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20northeast.%20Then%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EPark%20Lawn%20Road%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aphTQR1HfWw}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=16972, transferredBytes=0, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20600%20meters,%20Turn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aphTQR1HfWw}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=9135, transferredBytes=0, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right%20toward%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aphTQR1HfWw}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=12584, transferredBytes=0, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20300%20meters,%20Turn%20left.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aphTQR1HfWw}

----- (re-route to never driven route) -----
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=32802, transferredBytes=32802, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EDrive%20east%20on%20%3Csay-as%20interpret-as=%22address%22%3EThe%20Queensway%3C%2Fsay-as%3E.%20Then,%20in%20400%20meters,%20Turn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EStephen%20Drive%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aFNHc3fTtWy}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=18539, transferredBytes=18539, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20left%20onto%20%3Csay-as%20interpret-as=%22address%22%3EStephen%20Drive%3C%2Fsay-as%3E.%20Then%20Turn%20right.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aFNHc3fTtWy}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=18853, transferredBytes=18853, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3ETurn%20right.%20Then%20You%20will%20arrive%20at%20your%20destination.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aFNHc3fTtWy}
D/Mapbox: [nav-sdk]: [ResourceLoadObserver] onFinish: {status=AVAILABLE, totalBytes=11173, transferredBytes=0, url=https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EYou%20have%20arrived%20at%20your%20destination.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10aFNHc3fTtWy}

---- (route end) ------
```

<details><summary>Expand for a code that was used to capture above logs</summary>

```kotlin
object ResourceLoaderFactory {

    private val sharedLoader: DefaultResourceLoader by lazy {
        DefaultResourceLoader(
            TileStore.create(),
            ReachabilityFactory.reachability(null)
        ).apply {
            registerObserver(object : ResourceLoadObserver {
                override fun onStart(request: ResourceLoadRequest) = Unit

                override fun onProgress(
                    request: ResourceLoadRequest,
                    progress: ResourceLoadProgress
                ) = Unit

                override fun onFinish(
                    request: ResourceLoadRequest,
                    result: Expected<ResourceLoadError, ResourceLoadResult>
                ) {
                    result.value?.also { v ->
                        val cleanURL = request.url.replace(
                            Regex("access_token=([a-zA-Z0-9.]+)"),
                            "access_token=REDACTED"
                        )
                        val values = mutableMapOf(
                            "status" to v.status,
                            "totalBytes" to v.totalBytes,
                            "transferredBytes" to v.transferredBytes,
                            "url" to cleanURL
                        )
                        logD("onFinish: $values", "ResourceLoadObserver")
                    }
                }
            })
        }

    }

    /**
     * Returns default ResourceLoader.
     */
    fun getInstance(): ResourceLoader = sharedLoader
}
```

<details>

abhishek1508 commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechApi.kt:
```diff
@@ -78,71 +75,49 @@ class MapboxSpeechApi @JvmOverloads constructor(
+        if (VoiceInstructionsParser.parse(voiceInstruction).isError) {
+            throw IllegalStateException(
+                "Invalid state: processVoiceAnnouncement can't produce " +
+                    "Fallback VoiceTypeAndAnnouncement VoiceResult"
```
Error message needs to be changed since there is no action, result and processor anymore in this PR.

abhishek1508 replied:
Looks like `VoiceInstructionParser` returns the following `error` 
```
private fun invalidInstructionsError() =
        Error("VoiceInstructions ssmlAnnouncement / announcement can't be null or blank")
```
We can just propagate this error as `IllegalStateException`

tomaszrybakiewicz replied:
Done.
